File: android/app/src/main/java/com/mycampusdock/dock/FirebaseModule.java
package com.mycampusdock.dock;

import android.util.Log;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.google.firebase.messaging.FirebaseMessaging;

public class FirebaseModule extends ReactContextBaseJavaModule {
    public FirebaseModule(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return "FirebaseModule";
    }

    @ReactMethod
    public void subscribeTag(String tag) {
        FirebaseMessaging.getInstance().subscribeToTopic(tag);
        Log.d("Dock", "SUBSCRIBED : "+tag);
    }

    @ReactMethod
    public void unsubscribeTag(String tag) {
        FirebaseMessaging.getInstance().unsubscribeFromTopic(tag);
        Log.d("Dock", "UNSUBSCRIBED : "+tag);
    }

    @ReactMethod
    public void subscribeTags(ReadableArray tags) {
        for(int i=0; i<tags.size(); i++){
            subscribeTag(tags.getString(i));
        }
    }

    @ReactMethod
    public void unsubscribeTags(ReadableArray tags) {
        for(int i=0; i<tags.size(); i++){
            unsubscribeTag(tags.getString(i));
        }
    }

    @ReactMethod
    public void log(String log){
        Log.d("Dock JSLog", log);
    }
}

File: ios/Dock/FirebaseModule.m
#import "FirebaseModule.h"
#import <React/RCTLog.h>
@import Firebase;
@implementation FirebaseModule

RCT_EXPORT_MODULE();

+ (id)allocWithZone:(NSZone *)zone {
  static FirebaseModule *sharedInstance = nil;
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    sharedInstance = [super allocWithZone:zone];
  });
  return sharedInstance;
}

- (NSArray<NSString *> *)supportedEvents {
  return @[@"FCM_MSSG"];
}

-(void)tellJs:(NSString *)mssg
{
  [self sendEventWithName:@"FCM_MSSG" body:@{@"mssg": mssg}];
}

RCT_EXPORT_METHOD(log:(NSString *)mssg)
{
  RCTLogInfo(@"JSLOG %@", mssg);
}

RCT_EXPORT_METHOD(subscribeTag:(NSString *)tag)
{
  [[FIRMessaging messaging] subscribeToTopic:tag
                                  completion:^(NSError * _Nullable error) {
                                    NSLog(@"Subscribed to %@ topic", tag);
                                  }];
}

RCT_EXPORT_METHOD(unsubscribeTag:(NSString *)tag)
{
  [[FIRMessaging messaging] unsubscribeFromTopic:tag
                                  completion:^(NSError * _Nullable error) {
                                    NSLog(@"Unsubscribed to %@ topic", tag);
                                  }];
}

RCT_EXPORT_METHOD(subscribeTags:(NSArray *)tags)
{
  for (NSString* o in tags)
  {
    [self subscribeTag:o];
  }
}

RCT_EXPORT_METHOD(unsubscribeTags:(NSArray *)tags)
{
  for (NSString* o in tags)
  {
    [self unsubscribeTag:o];
  }
}

@end
